### PR TITLE
[SYCL][ESIMD] Make lsc_block_load/store try to use 64 bit blocks by default

### PR DESCRIPTION
**Pull request by fineg74**

### CODE REVIEW

**Review by sarnex — APPROVED** [review]

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:561]: Does this statement basically tells that lsc_block_load of simd<int, 2> with lsc_data_size::default_size, then the address 'p' must be aligned by 8? 
If so, then that seems wrong to me because for load of simd<int,2> still should be possible for address aligned by only 4.

Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:561]: IMO, using bigger element size (assuming it requires also bigger address alignment) would require extension of this API by adding hint/guarantee from user that the address passed to this routine is properly aligned.
'overaligned_tag<8>' could be used for that purpose.

**Review by v-klochkov — CHANGES_REQUESTED** [review]
Add 'Request changes' tag til resolution of the address alignment question.

**Review by v-klochkov — COMMENTED** [review]
At this moment I only reviewed it from API point of view. Please see some suggestions on adding non-template parameter to lsc_block_load/store.

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:616]: Gregory, I believe what we agreed on the default value previously was: std::max(__ESIMD_NS::overaligned_tag<4>, __ESIMD_NS::element_aligned_tag).

So, passing int64_t pointer would assume element_aligned_tag (i.e. 8-byte alignment) and would not require explicit passing of the tag argument for longs.

It is not possible to write std::max as shown above, but we can have something like 'dqword_element_aligned_tag' defined in _detail_ namespace, which would return 4 for byte/short/int and 8 for int64_t.
That is a temporary solution before we get approval for using element_aligned_tag that breaks backward compatibility with previous calls with int8_t* and int16_t* without explicit alignment tag passed.

Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:616]: Gregory, I believe what we agreed on the default value previously was: std::max(__ESIMD_NS::overaligned_tag<4>, __ESIMD_NS::element_aligned_tag).
So, passing int64_t pointer would assume element_aligned_tag (i.e. 8-byte alignment) and would not require explicit passing of the tag argument for longs.

From other side, if user explicitly passes overaligned<4> for 'int64_t* ptr', then that may be an indicator of
some alignment problems in users's code and it would be good to report that in compile time.
The latest update of this PR handles this scenario differently,
the explicitly set alignment to 4 would be just ignored.

It is not possible to write std::max as shown above, but we can have something like 'dqword_element_aligned_tag' defined in detail namespace, which would return 4 for byte/short/int and 8 for int64_t.
That also would remove the need in Alignment evaluation on the lines 625-632

That is a temporary solution before we get approval for using element_aligned_tag that breaks backward compatibility with previous calls with int8_t* and int16_t* without explicit alignment tag passed.

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:636]: If element type in int64_t, the required alignment is >= 8-bytes

File: sycl/include/sycl/ext/intel/esimd/detail/simd_obj_impl.hpp
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/esimd/detail/simd_obj_impl.hpp:79]: I think I have the last comment for this PR: Please move this class definition to detail namespace.
Using this class is rather a temporary internal solution used until we get approval for using element_aligned_tag for char/short loads.

**Review by v-klochkov — APPROVED** [review]

**Pull request merged**
